### PR TITLE
[PATCH v4] api: tm: split tm queue and node threshold capabilities

### PR DESCRIPTION
**Pull request by MatiasElo**
V3:
- Added validation tests for TM queue threshold

### CODE REVIEW

File: include/odp/api/spec/traffic_mngr.h [hunk 1]
```diff
@@ -1968,9 +1950,7 @@ int odp_tm_queue_disconnect(odp_tm_queue_t tm_queue);
  * after any non-zero shaper_len_adjust that is part of the shaper profile.
  *
  * The pkt_color bits are a result of some earlier Metering/Marking/Policing
- * processing (typically ingress based), and should not be confused with the
- * shaper_color produced from the TM shaper entities within the tm_inputs and
- * tm_nodes.
+ * processing (typically ingress based).
```
[the remark this answers is not in the document]
Comment (MatiasElo) [reply include/odp/api/spec/traffic_mngr.h:1974]: Ok, removed in V2.

File: include/odp/api/spec/traffic_mngr.h [hunk 2]
```diff
@@ -1941,10 +1941,11 @@ int odp_tm_queue_disconnect(odp_tm_queue_t tm_queue);
...
 
-/** The odp_tm_enq() function is used to add packets to a given TM system.
- * Note that the System Metadata associated with the pkt needed by the TM
- * system is (a) a drop_eligible bit, (b) a two bit "pkt_color", (c) a 16-bit
- * pkt_len, and MAYBE? (d) a signed 8-bit shaper_len_adjust.
+/** Send packet to TM system
+ *
+ * Note that the packet metadata utilized by the TM system is (a) a
+ * drop_eligible bit, (b) a two bit pkt_color, (c) a 16-bit pkt_len, and (d) a
+ * signed 8-bit shaper_len_adjust.
```
[the remark this answers is not in the document]
Comment (MatiasElo) [reply include/odp/api/spec/traffic_mngr.h:1968]: Fixed in V2.

File: include/odp/api/spec/traffic_mngr.h [hunk 3]
```diff
@@ -510,11 +510,18 @@ typedef struct {
 	/** TM queue statistics counter capabilities */
 	odp_tm_queue_stats_capability_t queue_stats;
 
-	/** tm_queue_threshold indicates support for threshold profile on a
-	 * TM queue. When TRUE, users can set/clear/update threshold profile
-	 * on a TM queue. When false none of it is supported.
-	 */
-	odp_bool_t tm_queue_threshold;
+	/** TM queue threshold profile support */
```
[the remark this answers is not in the document]
Comment (MatiasElo) [reply include/odp/api/spec/traffic_mngr.h:523]: Fixed in V2.

File: test/validation/api/traffic_mngr/traffic_mngr.c [hunk 4]
```diff
@@ -4563,7 +4564,8 @@ static int traffic_mngr_check_thresholds(void)
 	/* Check only for tm queue threshold support as
 	 * we only test queue threshold.
 	 */
-	if (!tm_capabilities.tm_queue_threshold)
+	if (!tm_capabilities.tm_queue_threshold.byte &&
+	    !tm_capabilities.tm_queue_threshold.packet)
```
[the remark this answers is not in the document]
Comment (MatiasElo) [reply test/validation/api/traffic_mngr/traffic_mngr.c:4569]: In this first version I did only the minimal test changes to get the code to compile. Once we agree on the API changes I'll add the test improvements, including this one.

File: include/odp/api/spec/traffic_mngr.h [hunk 5]
```diff
@@ -1753,8 +1753,13 @@ typedef struct {
 
 	/** The threshold profile to be used in setting the max queue fullness
 	 * for WRED and/or tail drop. Can be ODP_TM_INVALID and can also be set
-	 * and changed post-creation via odp_tm_queue_threshold_config(). The
-	 * default value is ODP_TM_INVALID. */
+	 * and changed post-creation via odp_tm_queue_threshold_config() if
+	 * 'dynamic_threshold_update' is supported. The default value is
```
[the remark this answers is not in the document]
Comment (MatiasElo) [reply include/odp/api/spec/traffic_mngr.h:1757]: True, dropped in V2.

File: test/validation/api/traffic_mngr/traffic_mngr.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (JannePeltonen) [reply test/validation/api/traffic_mngr/traffic_mngr.c:4569]: In that case I think this check here should be such that both byte and packet threshold must be supported for the test to be run. Now the capability check here is not consistent with what is being tested.

[the remark this answers is not in the document]
Comment (MatiasElo) [reply test/validation/api/traffic_mngr/traffic_mngr.c]: Fixed in V3.

**Review by JannePeltonen — APPROVED** [review]

**Review by psavol — APPROVED** [review]

**Comment by nithind1988** [comment]
Reviewed-by: Nithin Dabilpuram <ndabilpuram@marvell.com>

**Comment by nithind1988** [comment]
Looks good to me.

**Pull request merged**